jbmc/unit/java-testing-utils/require_goto_statements.cpp

```diff
@@ -295,6 +294,49 @@ const code_declt &require_goto_statements::require_declaration_of_name(
   throw no_decl_found_exceptiont(variable_name.c_str());
 }
 
+/// Follow the chain of non-null assignments until we find a symbol that
+/// hasn't ever had another symbol assigned to it.
+/// \param entry_point_instructions: A vector of instructions
+/// \param input_symbol_identifier: The identifier of the symbol we are
+///   currently considering
+/// \return The identifier of the symbol which is (possibly indirectly) assigned
+///   to \p input_symbol_identifier and which does not have any symbol assigned
+///   to it
+static const irep_idt &
+get_symbol_at_end_of_chain_of_non_null_assignments(
+  const std::vector<codet> &entry_point_instructions,
+  const irep_idt &input_symbol_identifier)
+{
+  // Looking at non-null assignments to input_symbol_identifier, there are two
+  // possible cases:
+  // 1. input_symbol_identifier has been dynamically allocated, so we'll find
+  //    code like this:
+  //      malloc_site$0 = ALLOCATE(struct java.lang.Object, 4ul, false);
+  // 2. input_symbol_identifier has another symbol assigned to it
+  //
+  // In practice it seems that case 2 is always followed by case 1, but this
+  // function has been written to deal with a slightly more general case.
+  const auto &non_null_assignments_to_input_symbol =
+    require_goto_statements::find_pointer_assignments(
+      input_symbol_identifier, entry_point_instructions)
+      .non_null_assignments;
+
+  REQUIRE(non_null_assignments_to_input_symbol.size() == 1);
+
+  const symbol_exprt *symbol_assigned_to_input_symbol =
+    expr_try_dynamic_cast<symbol_exprt>(
+      skip_typecast(non_null_assignments_to_input_symbol[0].rhs()));
+
+  // Check for case 2
```

Comment:
I think these case labels are the wrong way around? Case 2 is when there *is* a source symbol